feat: new extension `vike-react-chakra`

phonzammi opened the pull request:
TODO :

- [x] README

Add Custom Config/Setting :
~~ChakraCustomProvider ?~~
~~chakraCustomConfig ?~~
~~+system~~
```ts
chakra?: null | {
     system?: SystemContext
     locale?: LocaleProviderProps['locale']
}
```

brillout commented:
:100: 

Hi Muhammad, yes good point I think a new setting `+ChakraProvider` would be great. But let me know if you think an alternative idea could be better.

Looking forward to merge :heart_eyes:

phonzammi commented:
Hi Rom, Thanks
I just added `+system` setting and an example.

I'll also try adding the `+ChakraProvider` setting, which would allow users to use a custom provider, like adding `ThemeProvider` from [next-themes](https://github.com/pacocoursey/next-themes). However, I think users can also achieve this by using the +Wrapper setting.

brillout commented:
Hello Muhammad, nice :ok_hand:

I suggest we minimize the code a bit. The rationale being that we'll eventually have a *lot* of integration packages like this. To make things more maintainable let's keep things as simple as possible.

- Let's remove the example.
- Let's trim the readme a bit: let's assume the user already knows the basics of Chakra (but I ain't familiar with Chakra so I ain't sure what this means exactly).

> I just added `+system` setting

:+1: Is `system` the only prop of `ChakraProvider`? In other words: would replacing `+system` with `+ChakraProvider` make sense?

brillout commented:
> I'll also try adding the `+ChakraProvider` setting, which would allow users to use a custom provider, like adding `ThemeProvider` from [next-themes](https://github.com/pacocoursey/next-themes)

Ah, I'm just reading that :sweat_smile:. So, yea, I'm inclined to think that haivng a `+ChakraProvider` setting would be both simpler and more powerful than having a `+system` setting?

> I think users can also achieve this by using the `+Wrapper` setting.

Yea, maybe this integration package is a bit counter productive :thinking: WDYT?

Maybe with eject it could make sense?

phonzammi commented:
I'm not great at explaining this, but as far as I know, `ChakraProvider` only needs the theme system to work. The `next-theme` package isn’t necessary for it to function. However, the `@chakra-ui/cli` tool can adds code snippets that also installs the `next-theme` and `react-icons` packages, which are used in some of the pre-built component files generated by the CLI

brillout commented:
> However, the `@chakra-ui/cli` tool can adds code snippets that also installs the `next-theme` and `react-icons` packages, which are used in some of the pre-built component files generated by the CLI

How is this relevant to our discussion and/or to `vike-react-chakra`?

phonzammi commented:
So that users can follow the [Setup Provider](https://www.chakra-ui.com/docs/get-started/frameworks/vite#setup-provider) guide to use the generated Provider (which includes `ThemeProvider`) created by the CLI

phonzammi commented:
I just pushed a new changes.
Please take a look and review it

phonzammi commented:
I'll add `@brillout/release-me` package when we ready to merge it

brillout commented:
> So that users can follow the [Setup Provider](https://www.chakra-ui.com/docs/get-started/frameworks/vite#setup-provider) guide to use the generated Provider (which includes `ThemeProvider`) created by the CLI

Their docs say:

> This provider composes the following:
> - `ChakraProvider` from `@chakra-ui/react` for the styling system
> - `ThemeProvider` from `next-themes` for color mode

But that isn't something the user can influence, right?

In other words: what would be the concrete added value of `+ChakraProvider` over `+system`?

Alternative idea:

```js
// +chakra.js

export default {
  system: /* ... */
}
```

WDYT?

phonzammi commented:
> But that isn't something the user can influence, right?

It's what the CLI generates, (inside `./src/components/ui` dir), users can modify it as they want, or even create their own.

> In other words: what would be the concrete added value of `+ChakraProvider` over `+system`?

- If users use `+ChakraProvider.js` it's like overriding the `vike-react-chakra`'s `+Wrapper.js`
- `+system.js` only change/overriding the default chakra theme system configuration
- If users choose to create a new `+Wrapper.js` to add `ThemeProvider` and also want to be able to change the chakra theme system, we need to keep `+system` setting.

brillout commented:
> It's what the CLI generates, (inside `./src/components/ui` dir), users can modify it as they want, or even create their own.

I don't see how it's related to our discussion?

> * If users use `+ChakraProvider.js` it's like overriding the `vike-react-chakra`'s `+Wrapper.js`

Sure, but again, what's the concrete advantage?

> * `+system.js` only change/overriding the default chakra theme system configuration

Of course.

> * If users choose to create a new `+Wrapper.js` to add `ThemeProvider`

Does their docs document that? :eyes:

I'm still not sure I see any concrete benefit of `+ChakraProvider` over `+system`.

phonzammi commented:
> I don't see how it's related to our discussion?

Maybe I misunderstood your question. What do you mean by `something the user can influence` ?

phonzammi commented:
Anyways, what about we don't use `+ChakraProvider` for now and use `+system` setting with your alternative idea.
 
```js
// +chakra.js
 
export default {
   system: /* ... */
   locale: /* ... */
}
 ```

And we also can add optinal `locale?: string` too. Chakra UI has `<LocaleProvider>` component. So the final `+Wrapper.ts` file will look like this:

```ts
export { Wrapper }

import React, { ReactNode } from 'react'
import { ChakraProvider, LocaleProvider, defaultSystem } from '@chakra-ui/react'
import { usePageContext } from 'vike-react/usePageContext'

function Wrapper({ children }: { children: ReactNode }) {
  const pageContext = usePageContext()
  const { chakra } = pageContext.config

  return (
    <ChakraProvider value={chakra?.system ?? defaultSystem}>
      <ChakraLocaleProvider locale={chakra?.locale}>
        {children}
      </ChakraLocaleProvider>
    </ChakraProvider>
  )
}

function ChakraLocaleProvider(props: { locale?: string; children: ReactNode }) {
  if (props.locale) {
    return <LocaleProvider locale={props.locale}>{props.children}</LocaleProvider>
  }
  return <>{props.children}</>
}

```

phonzammi commented:
> * A default `system`? I don't know Chakra so I don't know if that can make sense.

I'm not sure either, though I think the `defaultSystem` from Chakra UI is good enough to start with.

Okay, I'm happy with this PR. Please take a look and let me know if I missed anything.

brillout commented on packages/vike-react-chakra/integration/+config.ts:
```diff
@@ -0,0 +1,31 @@
+export { config as default }
+
+import { LocaleProviderProps, SystemContext } from '@chakra-ui/react'
```
`import type { ... }`

brillout commented on packages/vike-react-chakra/integration/Wrapper.tsx:
```diff
@@ -0,0 +1,27 @@
+export { Wrapper }
+
+import React, { ReactNode } from 'react'
```
`type ReactNode`

brillout commented on packages/vike-react-chakra/integration/Wrapper.tsx:
```diff
@@ -0,0 +1,27 @@
+export { Wrapper }
+
+import React, { ReactNode } from 'react'
+import { ChakraProvider, LocaleProvider, defaultSystem } from '@chakra-ui/react'
```
`import { type ChakraProvider, type LocaleProvider, defaultSystem } from '@chakra-ui/react'`

brillout commented on packages/vike-react-chakra/integration/Wrapper.tsx:
```diff
@@ -0,0 +1,27 @@
+  )
+}
+
+function ChakraLocaleProvider({ locale, children }: { locale?: string; children: ReactNode }) {
```
Nice :sparkles:

brillout commented on packages/vike-react-chakra/package.json:
```diff
@@ -0,0 +1,51 @@
+{
+    "name": "vike-react-chakra",
+    "version": "0.0.1",
```
I suggest we go ahead and directly release `1.0.0`.

phonzammi replied:
Noted 👍

brillout commented on packages/vike-react-chakra/package.json:
```diff
@@ -0,0 +1,51 @@
+    "type": "module",
+    "exports": {
+        "./config": "./dist/integration/+config.js",
+        "./__internal/integration/Wrapper": "./dist/integration/Wrapper.js"
```
Given that `vike-react-chakra` has few files, let's remove the `integration/` directory. Let's also remove it in the export path.

brillout replied:
Also let's `mv Wrapper.tsx +Wrapper.tsx` so that it matches what we will eject (once we finished the eject implementation).

(FYI I'll update the `vike-{react,vue}` packages to also prefix all entries with `+`.)

phonzammi replied:
Then, we can remove this line too ?
```js
// +config.ts
// ...
  Wrapper: 'import:vike-react-chakra/__internal/integration/Wrapper:Wrapper',
// ...
```

brillout replied:
We still need it, Vike never crawls inside `node_modules/`.

phonzammi replied:
WDYT https://github.com/vikejs/vike-react/pull/151/commits/de3e81166390282e9fd5187d805badfd0a5d9459 ?

brillout replied:
Perfect :sparkles:

brillout commented on packages/vike-react-chakra/package.json:
```diff
@@ -0,0 +1,51 @@
+        "release:minor": "release-me minor",
+        "release:commit": "release-me commit"
+    },
+    "peerDependencies": {
```
Nice optimistic range, good move :sparkles:

brillout commented on packages/vike-react-chakra/package.json:
```diff
@@ -0,0 +1,51 @@
+        "vike-react": ">=0.4.13"
+    },
+    "devDependencies": {
+        "@brillout/release-me": "0.3.8",
```
`^0.3.8`

brillout commented on packages/vike-react-chakra/package.json:
```diff
@@ -0,0 +1,51 @@
+        "@emotion/react": "^11.13.3",
+        "@types/node": "^20.11.17",
+        "@types/react": "^18.2.55",
+        "jsdom": "^24.0.0",
```
That's surprising. Is it a peer dep? :thinking:

FYI the whole `vike-react-chakra` package will soon be ejectable, so polishing is extra worth it.

phonzammi replied:
`jsdom` ?, my bad, I'll remove it. As for `@types/node`, let me check if we need it here

brillout commented on packages/vike-react-chakra/package.json:
```diff
@@ -0,0 +1,51 @@
+    "files": [
+        "dist"
+    ],
+    "repository": "github:vikejs/vike-react",
```
Let's update this to be a more precise URL.

brillout replied:
AFAICT `"repository": "https://github.com/vikejs/vike-react/tree/main/packages/vike-react-chakra"` is enough.

brillout commented on packages/vike-react-chakra/tsconfig.json:
```diff
@@ -0,0 +1,24 @@
+{
```
Let's try to minimize this to the strict minimum while keeping `"sctrict": true`.

We'll copy-paste this a lot for future `vike-react-*` packages, so it's worth it to minimize it.

phonzammi replied:
Is it okay like this https://github.com/vikejs/vike-react/pull/151/commits/6f58d8e904a30768a4f1fa2f68d2366edd8b9fa7 ?

brillout commented on packages/vike-react-chakra/README.md:
```diff
@@ -0,0 +1,97 @@
+<!-- WARNING: keep links absolute in this file so they work on NPM too -->
```
To increase maintainability, let's further minimize the Readme. Let's remove the Vike logo.

brillout commented on packages/vike-react-chakra/README.md:
```diff
@@ -0,0 +1,97 @@
+
+## Installation
+
+1. npm install @chakra-ui/react @emotion/react vike-react-chakra
```
Backticks are missing. Let's put `vike-react-chakra` first in the install list.

brillout commented on packages/vike-react-chakra/README.md:
```diff
@@ -0,0 +1,97 @@
+
+<br/>
+
+## Basic usage
```
Maybe we can remove the whole `Basic usage` section with something like the following. WDYT?

> That's it, you can now use Chakra components for any of your component.
>
> ```jsx
> import { HStack, Button } from '@chakra-ui/react'
> const Demo = () => {
>   return (
>     <HStack>
>       <Button>Click me</Button>
>       <Button>Click me</Button>
>     </HStack>
>   )
> }
> ```

brillout replied:
So essentially removing the note and the heading.

brillout commented on packages/vike-react-chakra/README.md:
```diff
@@ -0,0 +1,97 @@
+<br/>
+
+## `+chakra` Setting
+`vike-react-chakra` provides a custom setting, `+chakra`, for configuring custom theme system, custom locale to adds `LocaleProvider` and can be used to disable Chakra UI for specific pages.
```
```suggestion
`vike-react-chakra` provides a configuration `+chakra` for setting a custom theme system and custom locale.
```

Let's mention `can be used to disable Chakra UI for specific pages` later.

brillout commented on packages/vike-react-chakra/README.md:
````diff
@@ -0,0 +1,97 @@
+## `+chakra` Setting
+`vike-react-chakra` provides a custom setting, `+chakra`, for configuring custom theme system, custom locale to adds `LocaleProvider` and can be used to disable Chakra UI for specific pages.
+```js
+// pages/+chakra.js
````
```suggestion
// pages/+chakra.js

```

brillout commented on packages/vike-react-chakra/README.md:
````diff
@@ -0,0 +1,97 @@
+`vike-react-chakra` provides a custom setting, `+chakra`, for configuring custom theme system, custom locale to adds `LocaleProvider` and can be used to disable Chakra UI for specific pages.
+```js
+// pages/+chakra.js
+import React, { ReactNode } from 'react'
````
```suggestion
import React, { type ReactNode } from 'react'
```

brillout commented on packages/vike-react-chakra/README.md:
````diff
@@ -0,0 +1,97 @@
+}
+```
+
+To disable Chakra UI for specific page:
````
```suggestion
You can remove Chakra UI [for some of your pages](https://vike.dev/config#inheritance):
```

brillout commented on packages/vike-react-chakra/README.md:
````diff
@@ -0,0 +1,97 @@
+
+To disable Chakra UI for specific page:
+```js
+// pages/about/+chakra.js
````
```suggestion
// pages/about/+chakra.js

```

brillout commented on packages/vike-react-chakra/README.md:
```diff
@@ -0,0 +1,97 @@
+
+<br/>
+
+## See also
```
Good idea but I think it's better we remove it to keep things maintainable; we will have a *lot* of `vike-react-*` packages.

brillout commented on packages/vike-react-chakra/README.md:
```diff
@@ -0,0 +1,97 @@
+
+export const chakra = {
+  system,
+  locale: "ar-BH"
```
Let's showcase one that is more widely known e.g. `fr-FR`.

brillout commented on packages/vike-react-chakra/README.md:
```diff
@@ -0,0 +1,97 @@
+
+<br/>
+
+## `+chakra` Setting
```
```suggestion
## `+chakra` Setting

```

brillout commented:
@phonzammi Let me know once you're happy with the PR and I'll have another (probably final) look at it.

phonzammi commented:
@brillout, Thank you. I'm happy with this PR. Looking forward for another reviews from you. 

Also looking forward to create another Vike extensions 🚀

brillout commented:
I made a couple of commits, let me know whether you agree with all of it and I'll merge.

> Also looking forward to create another Vike extensions 🚀

Sounds good :100: and let's progressively minimize the number of interventions needed from my side.

brillout commented:
Btw. I changed my mind and removed `+` because having `+` files in `ejected/vike-react-chakra/` is actually wrong (Vike shouldn't crawl these).

phonzammi commented:
> Btw. I changed my mind and removed `+` because having `+` files in `ejected/vike-react-chakra/` is actually wrong/misleading (Vike shouldn't crawl these).

I think we still need to use `+config.js` file or we get this errors:
```js

9:22:21 PM [vike][config][Wrong Usage] /pages/+config.ts sets the config extends to an invalid value, see https://vike.dev/extends
9:22:21 PM [vike][Warning] Update to the new V1 design to get rid of Vite's warning: The glob option "as" has been deprecated in favour of "query". See https://vike.dev/migration/v1-design for how to migrate.
The glob option "as" has been deprecated in favour of "query". Please update `as: 'extractExportNames'` to `query: '?extractExportNames'`.
```

phonzammi commented:
> if you still have the example somewhere ...

Sorry, I don't have any example on my repo yet

brillout commented:
It now works with `vike@0.4.203`.

phonzammi commented:
Well done, awesome 🚀 , thank you

phonzammi commented:
LGTM, even better we don't need to use `+` for `config.js`. 

FYI, I had issues with `vike-solid-query` because of that, sorry I didn't ask you before

brillout commented:
Merged :sparkles: 

Let's:
 - Release it
 - Add it to vike.dev/extensions
 - Create vike.dev/chakra

brillout commented:
Also we forgot to update the main readme: https://github.com/vikejs/vike-react

brillout commented:
> Also we forgot to update the main readme: https://github.com/vikejs/vike-react

Or better yet let's remove that readme and just link it to vike.dev/extensions#react

"Monorepo of all [oficial Vike React extensions](...)"

phonzammi commented:
> Or better yet let's remove that readme and just add link to vike.dev/extensions#react

I agree

brillout commented:
> Let's:
> 
> * Release it
> * Add it to vike.dev/extensions
> * Create vike.dev/chakra

> > Also we forgot to update the main readme: https://github.com/vikejs/vike-react
> 
> Or better yet let's remove that readme and just add link to vike.dev/extensions#react
> 
> `Monorepo of [Vike React extensions](...)`

@phonzammi Up for it?

phonzammi commented:
Sure, I want to, ETA later tonight
